File: [I]stack-queue/[I]stack-queue-monotonic/739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        res = [0 for _ in range(len(temperatures))]
        stack = []
        for i in range(len(temperatures)):
            while stack and temperatures[stack[- 1]] < temperatures[i]:
                idx = stack.pop()
                res[idx] = i - idx
            stack.append(i)
        return res
    
# time O(n), each temperature will only pop and push once and both cost O(1)
# space O(n), due to stack's size
# using stack and queue and montonic and monotonic stack (consider one side’s relationship)

File: [I]stack-queue/[I]stack-queue-monotonic/739-daily-temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int[] res = new int[temperatures.length];
        ArrayDeque<Integer> stack = new ArrayDeque<>();
        for (int i = 0; i < temperatures.length; i++) {
            while (!stack.isEmpty() && temperatures[stack.peek()] < temperatures[i]) {
                int idx = stack.pop();
                res[idx] = i - idx;
            }
            stack.push(i);
        }
        return res;
    }
}

// time O(n), each temperature will only pop and push once and both cost O(1)
// space O(n), due to stack's size
// using stack and queue and montonic and monotonic stack (consider one side’s relationship)